lib/queue_classic_plus/worker.rb

```diff
@@ -33,54 +33,55 @@ def handle_failure(job, e)
         force_retry = true
       end
 
-      klass = job_klass(job)
+      @job = job
 
-      if force_retry && !(klass.respond_to?(:disable_retries) && klass.disable_retries)
+      if force_retry && !(job_class.respond_to?(:disable_retries) && job_class.disable_retries)
         Metrics.increment("qc.force_retry", source: @q_name)
-
-        retry_with_remaining(klass, job, e, 0)
+        retry_with_remaining(e)
       # The mailers doesn't have a retries_on?
-      elsif klass && klass.respond_to?(:retries_on?) && klass.retries_on?(e)
+      elsif job_class && job_class.respond_to?(:retries_on?) && job_class.retries_on?(e)
         Metrics.increment("qc.retry", source: @q_name)
-
-        backoff = (max_retries(klass) - remaining_retries(klass, job)) * BACKOFF_WIDTH
-
-        retry_with_remaining(klass, job, e, backoff)
+        retry_with_remaining(e)
       else
-        enqueue_failed(job, e)
+        enqueue_failed(e)
       end
 
-      FailedQueue.delete(job[:id])
+      FailedQueue.delete(@job[:id])
     end
 
     private
 
-    def retry_with_remaining(klass, job, e, backoff)
-      @remaining_retries = remaining_retries(klass, job)
+    def retry_with_remaining(e)
+      @remaining_retries = calculate_remaining_retries
 
       if @remaining_retries > 0
-        klass.restart_in(backoff, @remaining_retries, *job[:args])
+        @remaining_retries -= 1
+        job_class.restart_in(backoff, @remaining_retries, *@job[:args])
       else
-        enqueue_failed(job, e)
+        enqueue_failed(e)
       end
     end
 
-    def max_retries(klass)
-      klass.respond_to?(:max_retries) ? klass.max_retries : 5
+    def max_retries
+      job_class.respond_to?(:max_retries) ? job_class.max_retries : 5
     end
 
-    def remaining_retries(klass, job)
-      @remaining_retries ? @remaining_retries - 1 : (job[:remaining_retries] || max_retries(klass)).to_i - 1
```

Comment:
So at some point when `@remaining_retries` wasn't yet set (L59), this line evaluated to `nil.to_i - 1` and we failed immediately?

Please confirm whether this is correct, I'd love to understand what went wrong

Reply:
Nope, the actual bug was caused by https://github.com/rainforestapp/queue_classic_plus/pull/31/files#diff-d65a4e711b781d02a6fce3e8272bb4cfL59 where we premeturely decrease the retry count before we actually call https://github.com/rainforestapp/queue_classic_plus/pull/31/files#diff-d65a4e711b781d02a6fce3e8272bb4cfL62.